File: data_eng/02_Spark/scala/src/main/scala/spark/etl02_LoadFIle.scala
package spark

import org.apache.spark.sql._
import org.apache.log4j._
import org.apache.spark.sql.types.{FloatType, IntegerType, StringType, StructType, BooleanType}

object etl02_LoadFIle {
  //Create Class
  case class inf_games( Title: String,
                        Features_Handheld: Boolean,
                        Features_Max_Players: Int,
                        Features_Multiplatform: Boolean,
                        Features_Online: Boolean,
                        Metadata_Genres: String,
                        Metadata_Licensed: Boolean,
                        Metadata_Publishers: String,
                        Metadata_Sequel: Boolean,
                        Metrics_Review_Score: Int,
                        Metrics_Sales: Float,
                        Metrics_Used_Price: Float,
                        Release_Console: String,
                        Release_Rating: String,
                        Release_Re_release: Boolean,
                        Release_Year: Int,
                        Length_All_PlayStyles_Average: Float,
                        Length_All_PlayStyles_Leisure: Float,
                        Length_All_PlayStyles_Median: Float,
                        Length_All_PlayStyles_Polled: Int,
                        Length_All_PlayStyles_Rushed: Float,
                        Length_Completionists_Average: Float,
                        Length_Completionists_Leisure: Float,
                        Length_Completionists_Median: Float,
                        Length_Completionists_Polled: Int,
                        Length_Completionists_Rushed: Float,
                        Length_Main_Extras_Average: Float,
                        Length_Main_Extras_Leisure: Float,
                        Length_Main_Extras_Median: Float,
                        Length_Main_Extras_Polled: Int,
                        Length_Main_Extras_Rushed: Float,
                        Length_Main_Story_Average: Float,
                        Length_Main_Story_Leisure: Float,
                        Length_Main_Story_Median: Float,
                        Length_Main_Story_Polled: Int,
                        Length_Main_Story_Rushed: Float)
  def main(args: Array[String]): Unit ={
    // Set the log level to only print errors
    Logger.getLogger("org").setLevel(Level.ERROR)
    //Use SparkSession
    val spark = SparkSession
      .builder
      .appName("SparkSQL")
      .master("local")
      .getOrCreate()

    //DataType in variable
    val gameSchema = new StructType()
      .add("Title", StringType, nullable = true)
      .add("Features_Handheld", BooleanType, nullable = true)
      .add("Features_Max_Players", IntegerType, nullable = true)
      .add("Features_Multiplatform", BooleanType, nullable = true)
      .add("Features_Online", BooleanType, nullable = true)
      .add("Metadata_Genres", StringType, nullable = true)
      .add("Metadata_Licensed", BooleanType, nullable = true)
      .add("Metadata_Publishers", StringType, nullable = true)
      .add("Metadata_Sequel", BooleanType, nullable = true)
      .add("Metrics_Review_Score", IntegerType, nullable = true)
      .add("Metrics_Sales", FloatType, nullable = true)
      .add("Metrics_Used_Price", FloatType, nullable = true)
      .add("Release_Console", StringType, nullable = true)
      .add("Release_Rating", StringType, nullable = true)
      .add("Release_Re_release", BooleanType, nullable = true)
      .add("Release_Year",  IntegerType, nullable = true)
      .add("Length_All_PlayStyles_Average", FloatType, nullable = true)
      .add("Length_All_PlayStyles_Leisure", FloatType, nullable = true)
      .add("Length_All_PlayStyles_Median", FloatType, nullable = true)
      .add("Length_All_PlayStyles_Polled", IntegerType, nullable = true)
      .add("Length_All_PlayStyles_Rushed", FloatType, nullable = true)
      .add("Length_Completionists_Average", FloatType, nullable = true)
      .add("Length_Completionists_Leisure", FloatType, nullable = true)
      .add("Length_Completionists_Median", FloatType, nullable = true)
      .add("Length_Completionists_Polled", IntegerType, nullable = true)
      .add("Length_Completionists_Rushed", FloatType, nullable = true)
      .add("Length_Main_Extras_Average", FloatType, nullable = true)
      .add("Length_Main_Extras_Leisure", FloatType, nullable = true)
      .add("Length_Main_Extras_Median", FloatType, nullable = true)
      .add("Length_Main_Extras_Polled", IntegerType, nullable = true)
      .add("Length_Main_Extras_Rushed", FloatType, nullable = true)
      .add("Length_Main_Story_Average", FloatType, nullable = true)
      .add("Length_Main_Story_Leisure", FloatType, nullable = true)
      .add("Length_Main_Story_Median", FloatType, nullable = true)
      .add("Length_Main_Story_Polled", IntegerType, nullable = true)
      .add("Length_Main_Story_Rushed", FloatType, nullable = true)

    //file .csv
    import spark.implicits._
    val arq = spark.read
      .schema(gameSchema)
      .option("header", "true")
      .csv("01_data/video_games.csv")
      .as[inf_games]

  try{
    // carregando a tabela no banco
    arq.write.format("jdbc")
      .option("url", "jdbc:mysql://localhost:3306/mydesenv")
      .option("dbtable","tb_bi_vgame")
      .option("user", "root")
      .option("password","mysql")
      .mode("Append")
      .save()
    }catch{
      case error: Throwable => println("banco mysql indisponivel",error)
    }
  }
}

File: data_eng/02_Spark/python/etl02_LoadFile.py
import datetime as dt
from pyspark.sql import SparkSession
from pyspark.sql.functions import *
from pyspark.sql.types import *


# conectando mysql
spark = SparkSession.builder \
        .master('local') \
        .appName("Conection Mysql") \
        .config("spark.driver.extraClassPath", "/Users/eduardoalberto/opt/spark/jars/mysql-connector-java-8.0.29.jar") \
        .getOrCreate()

# diretorio file
dirfiler = "/Users/eduardoalberto/LoadFile/video_games.csv"

def main():
    spark.sparkContext.setLogLevel("OFF")
    inf_games = StructType([
         StructField("Title"                        , StringType(), True)
        ,StructField("Features_Handheld"            , BooleanType(), True)
        ,StructField("Features_Max_Players"         , IntegerType(), True)
        ,StructField("Features_Multiplatform"       , BooleanType(), True)
        ,StructField("Features_Online"              , BooleanType(), True)
        ,StructField("Metadata_Genres"              , StringType(), True)
        ,StructField("Metadata_Licensed"            , BooleanType(), True)
        ,StructField("Metadata_Publishers"          , StringType(), True)
        ,StructField("Metadata_Sequel"              , BooleanType(), True)
        ,StructField("Metrics_Review_Score"         , IntegerType(), True)
        ,StructField("Metrics_Sales"                , FloatType(), True)
        ,StructField("Metrics_Used_Price"           , FloatType(), True)
        ,StructField("Release_Console"              , StringType(), True)
        ,StructField("Release_Rating"               , StringType(), True)
        ,StructField("Release_Re_release"           , BooleanType(), True)
        ,StructField("Release_Year"                 , IntegerType(), True)
        ,StructField("Length_All_PlayStyles_Average", FloatType(), True)
        ,StructField("Length_All_PlayStyles_Leisure", FloatType(), True)
        ,StructField("Length_All_PlayStyles_Median" , FloatType(), True)
        ,StructField("Length_All_PlayStyles_Polled" , IntegerType(), True)
        ,StructField("Length_All_PlayStyles_Rushed" , FloatType(), True)
        ,StructField("Length_Completionists_Average", FloatType(), True)
        ,StructField("Length_Completionists_Leisure", FloatType(), True)
        ,StructField("Length_Completionists_Median" , FloatType(), True)
        ,StructField("Length_Completionists_Polled" , IntegerType(), True)
        ,StructField("Length_Completionists_Rushed" , FloatType(), True)
        ,StructField("Length_Main_Extras_Average"   , FloatType(), True)
        ,StructField("Length_Main_Extras_Leisure"   , FloatType(), True)
        ,StructField("Length_Main_Extras_Median"    , FloatType(), True)
        ,StructField("Length_Main_Extras_Polled"    , IntegerType(), True)
        ,StructField("Length_Main_Extras_Rushed"    , FloatType(), True)
        ,StructField("Length_Main_Story_Average"    , FloatType(), True)
        ,StructField("Length_Main_Story_Leisure"    , FloatType(), True)
        ,StructField("Length_Main_Story_Median"     , FloatType(), True)
        ,StructField("Length_Main_Story_Polled"     , IntegerType(), True)
        ,StructField("Length_Main_Story_Rushed"     , FloatType(), True)
        ,StructField("dt_atualizacao"               , DateType(), True)
        ])
    arq = spark.read.csv(dirfiler, header=True, schema=inf_games)
    # add novo campo 
    tp = arq.withColumn("dt_atualizacao", lit(dt.datetime.now()))
    view = tp.select('Title','Features_Max_Players','Features_Online','Features_Multiplatform','Metadata_Genres','Metadata_Publishers','Metrics_Review_Score','Metrics_Sales','Metrics_Used_Price','Release_Year','Release_Rating','dt_atualizacao').show()
    try:
            view.write.format("jdbc").options(
            url="jdbc:mysql://localhost:3306/mydesenv",
            driver = "com.mysql.cj.jdbc.Driver",
            dbtable = "stg_video_game",
            user="root",
            password="mysql").mode('append').save()
    except:
            print('Erro ao tenta carregar banco !')

if __name__ == "__main__":
        main()

